src/Analyzer/ColumnNode.cpp

```diff
@@ -73,7 +73,9 @@ void ColumnNode::dumpTreeImpl(WriteBuffer & buffer, FormatState & state, size_t
 bool ColumnNode::isEqualImpl(const IQueryTreeNode & rhs) const
 {
     const auto & rhs_typed = assert_cast<const ColumnNode &>(rhs);
-    return column == rhs_typed.column;
+    auto src = getColumnSourceOrNull();
+    auto rhs_src = rhs_typed.getColumnSourceOrNull();
+    return column == rhs_typed.column && (src && rhs_src && src->isEqual(*rhs_src));
```

Comment:
It should not be done manually, `isEqualImpl` need to compare only internal state.
From code documentation:
```
/** Subclass must compare its internal state with rhs node internal state and do not compare children or weak pointers to other
  * query tree nodes.
  */
virtual bool isEqualImpl(const IQueryTreeNode & rhs) const = 0;
```
This was implemented that way, because it is possible to have recursive cycles during comparison of weak pointers.
For example column node is inside lambda, and lambda is column node source.

It is done properly in IQueryTreeNode.h `isEqual` method.

Reply:
Maybe we can just always compare aliases for TableNode? Because when we have `table` and `table as t1` in query tree normally it indicates that it's a different reads from the same table

Reply:
I suppose we don't need to call `isEqual` for column sources. We can just compare where these pointers refer to.